tools/trace2syz/trace2syz/intermediateTypes.go

```diff
@@ -0,0 +1,686 @@
+}
+
+//NewTraceTree initializes a TraceTree
+func NewTraceTree() (tree *TraceTree) {
+	tree = &TraceTree{
+		TraceMap: make(map[int64]*Trace),
+		Ptree:    make(map[int64][]int64),
+		RootPid:  -1,
+	}
+	return
+}
+
+func (tree *TraceTree) Contains(pid int64) bool {
+	if _, ok := tree.TraceMap[pid]; ok {
+		return true
+	}
+	return false
+}
+
+func (tree *TraceTree) Add(call *Syscall) *Syscall {
+	if tree.RootPid < 0 {
+		tree.RootPid = call.Pid
+	}
+	if !call.Resumed {
+		if !tree.Contains(call.Pid) {
+			tree.TraceMap[call.Pid] = newTrace()
+			tree.Ptree[call.Pid] = make([]int64, 0)
+		}
+	}
+	c := tree.TraceMap[call.Pid].Add(call)
+	if c.CallName == "clone" && !c.Paused {
+		tree.Ptree[c.Pid] = append(tree.Ptree[c.Pid], c.Ret)
+	}
+	return c
+}
+
+func (tree *TraceTree) String() string {
+	var buf bytes.Buffer
+
+	buf.WriteString(fmt.Sprintf("Root: %d\n", tree.RootPid))
+	buf.WriteString(fmt.Sprintf("Pids: %d\n", len(tree.TraceMap)))
+	return buf.String()
+}
+
+//Trace is just a list of system calls
+type Trace struct {
+	Calls []*Syscall
+}
+
+//newTrace initializes a new trace
+func newTrace() (trace *Trace) {
+	trace = &Trace{Calls: make([]*Syscall, 0)}
+	return
+}
+
+func (trace *Trace) Add(call *Syscall) (ret *Syscall) {
+	if call.Resumed {
+		lastCall := trace.Calls[len(trace.Calls)-1]
+		lastCall.Args = append(lastCall.Args, call.Args...)
+		lastCall.Paused = false
+		lastCall.Ret = call.Ret
+		ret = lastCall
+	} else {
+		trace.Calls = append(trace.Calls, call)
+		ret = call
+	}
+	return
+}
+
+//Syscall struct is the IR type for any system call
+type Syscall struct {
+	CallName string
+	Args     []irType
+	Pid      int64
+	Ret      int64
+	Cover    []uint64
+	Paused   bool
+	Resumed  bool
+}
+
+//NewSyscall - constructor
+func NewSyscall(pid int64, name string, args []irType, ret int64, paused bool, resumed bool) (sys *Syscall) {
+	sys = new(Syscall)
+	sys.CallName = name
+	sys.Args = args
+	sys.Pid = pid
+	sys.Ret = ret
+	sys.Paused = paused
+	sys.Resumed = resumed
+	return
+}
+
+//String
+func (s *Syscall) String() string {
+	var buf bytes.Buffer
+
+	buf.WriteString(fmt.Sprintf("Pid: %d-", s.Pid))
+	buf.WriteString(fmt.Sprintf("Name: %s-", s.CallName))
+	for _, typ := range s.Args {
+		buf.WriteString("-")
+		buf.WriteString(typ.String())
+		buf.WriteString("-")
+	}
+	buf.WriteString(fmt.Sprintf("-Ret: %d\n", s.Ret))
+	return buf.String()
+}
+
+type irType interface {
+	Name() string
+	String() string
+	Eval(*prog.Target) uint64
+}
+
+type dynamicType struct {
+	BeforeCall *expression
+	AfterCall  *expression
+}
+
+func newDynamicType(before, after irType) *dynamicType {
+	return &dynamicType{BeforeCall: before.(*expression), AfterCall: after.(*expression)}
+}
+
+func (d *dynamicType) String() string {
+	return d.BeforeCall.String()
+}
+
+func (d *dynamicType) Name() string {
+	return "Dynamic Type"
+}
+
+func (d *dynamicType) Eval(target *prog.Target) uint64 {
+	panic("Eval called on DynamicType")
+}
+
+type expression struct {
```

Comment:
it seems this needs to be an interface, it will get rid of all of the switches below

Reply:
ack